File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* node) 
    {
        vector<vector<int>> v;
        if(node == NULL)
        return v;

        queue<TreeNode*> q;
        q.push(node);
     
        bool flag=true;
        while(!q.empty())
        {
            int level =q.size();
            vector<int> temp(level);
            for(int i=0;i<level;i++)
            {
                TreeNode* curr=q.front();
                q.pop();
            
                if(flag)
                temp[i]=curr->val;
                else
                temp[level-i-1]=curr->val;
            
                if(curr->left!=NULL)
                q.push(curr->left);

                if(curr->right!=NULL)
                q.push(curr->right);
            }
            flag=!flag;
            v.push_back(temp);
        }
        return v;
    }
};

File: 103-binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        List<List<Integer>>  result = new ArrayList<>();
        if (root == null) return result;
        Queue<TreeNode> queue = new LinkedList<>();
        queue.add(root);
        
        boolean flag = false;
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            List<Integer> levelList = new ArrayList<>();
            for(int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                levelList.add(node.val);

                if(node.left != null) queue.add(node.left);
                if(node.right != null) queue.add(node.right);
            }
            if(flag) {
                Collections.reverse(levelList);
            }
            flag = !flag;
            result.add(new ArrayList<>(levelList));
        }
        return result;
    }
}